extensions/wikia/FilePage/FilePageHelper.class.php

```diff
@@ -67,10 +70,66 @@ public static function getFilePageRedirect( Title $title ) {
 				'file' => $title->getText()
 			] );
 		}
-		$wgMemc->set( $redirKey, $url );
+		if( !$onlyDB ) {
+			$wgMemc->set( $redirKey, $url );
+		}
 		return $url;
 	}
 
+	/**
+	 * Returns list of surrogate keys for purging.
+	 *
+	 * @requestParam Title
+	 *
+	 * @return array $keys - surrogate keys to purge
+	 */
+	public static function getSurrogateKeys( Title $title ) {
+		$keys = [];
+		$keys[] = self::getRedirSurrogateKey( $title );
+		if ( $title->inNamespace( NS_FILE ) ) {
+			$url = self::getFilePageRedirect( $title );
+			if( isset($url) ){
+				$pageTitle = self::rawURLToTitle( $url );
+				if( $pageTitle ){
+					$keys = array_merge( $keys, self::getSurrogateKeys( $pageTitle ) );
+				}
+			}
+			$url2 = self::getFilePageRedirect( $title, true );
+			if( isset($url2) && $url != $url2 ){
+				$pageTitle = self::rawURLToTitle( $url2 );
+				if( $pageTitle ){
+					$keys = array_merge( $keys, self::getSurrogateKeys( $pageTitle ) );
+				}
+			}
+		}
+		return $keys;
+	}
+
+	/**
+	 * Returns surrogate keys for purging
+	 *
+	 * @requestParam Title
+	 *
+	 * @return string $key - surrogate key to purge
+	 */
+	public static function getRedirSurrogateKey( Title $title ) {
+		return 'redirect-' . $title->getPrefixedText();
+	}
+
+	/**
+	 * Returns Title object based on araw URL
+	 *
+	 * @requestParam string
+	 *
+	 * @return Title $title or null
+	 */
+	private static function rawURLToTitle( string $url ) {
```

Comment:
Just a thought after analyzing the flow. maybe it would be easier to just store page title text in the memcache instead of the full url? This way we would avoid doing this magic here.